c2rust-refactor/src/scripting/ast_visitor.rs

```diff
@@ -0,0 +1,655 @@
+
+        Ok(ast)
+    }
+}
+
+impl<'lua> IntoLuaAst<'lua> for P<Item> {
+    fn into_lua_ast(self, ctx: &TransformCtxt, lua_ctx: LuaContext<'lua>) -> LuaResult<LuaTable<'lua>> {
+        let ast = lua_ctx.create_table()?;
+
+        ast.set("type", "Item")?;
+
+        self.and_then(|item| {
+            ast.set("id", item.id.as_u32())?;
+            ast.set("ident", item.ident.name.as_str().get())?;
+
+            match item.node {
+                ItemKind::ExternCrate(opt_name) => {
+                    ast.set("kind", "ExternCrate")?;
+
+                    if let Some(name) = opt_name {
+                        ast.set("name", name.as_str().get())?;
+                    }
+                },
+                ItemKind::Mod(module) => {
+                    ast.set("kind", "Mod")?;
+                    ast.set("mod", module.into_lua_ast(ctx, lua_ctx)?)?;
+                },
+                ItemKind::Use(_use_tree) => {
+                    ast.set("kind", "Use")?;
+                    // ast.set("use_tree", use_tree.into_lua_ast(ctx, lua_ctx))?;
+                },
+                ItemKind::Fn(decl, _header, _generics, block) => {
+                    ast.set("kind", "Fn")?;
+                    ast.set("decl", decl.into_lua_ast(ctx, lua_ctx)?)?;
+                    ast.set("block", block.into_lua_ast(ctx, lua_ctx)?)?;
+                },
+                ItemKind::Struct(_variant_data, _generics) => {
+                    ast.set("kind", "Struct")?;
+                    // TODO: Variant data, generics
+                },
+                ItemKind::Impl(.., items) => {
+                    let items = items
+                        .into_iter()
+                        .map(|item| item.into_lua_ast(ctx, lua_ctx))
+                        .collect::<LuaResult<Vec<_>>>();
+
+                    ast.set("kind", "Impl")?;
+                    ast.set("items", iter_to_lua_array(items?.into_iter(), lua_ctx)?)?;
+
+                    // TODO: other fields
+                },
+                ref e => warn!("Item IntoLuaAst kind: {:?}", e),
+            }
+
+            Ok(ast)
+        })
+
+    }
+}
+
+impl<'lua> IntoLuaAst<'lua> for ImplItem {
+    fn into_lua_ast(self, ctx: &TransformCtxt, lua_ctx: LuaContext<'lua>) -> LuaResult<LuaTable<'lua>> {
+        let ast = lua_ctx.create_table()?;
+
+        ast.set("type", "ImplItem")?;
+
+        match self.node {
+            ImplItemKind::Method(sig, block) => {
+                ast.set("kind", "ImplMethod")?;
+                ast.set("ident", self.ident.to_string())?;
+                ast.set("decl", sig.decl.into_lua_ast(ctx, lua_ctx)?)?;
+                ast.set("block", block.into_lua_ast(ctx, lua_ctx)?)?;
+                // TODO: generics, attrs, ..
+            },
+            ref e => unimplemented!("IntoLuaAst for {:?}", e),
+        }
+
+        Ok(ast)
+    }
+}
+
+pub(crate) trait MergeLuaAst {
+    fn merge_lua_ast<'lua>(self, table: LuaTable<'lua>) -> LuaResult<()>;
```

Comment:
Can we use `&mut self` here instead? The function shouldn't discard `self` (which is what this signature implies).

Reply:
I thought there was a reason I did it this way and impl'd for `&mut T`, but I can't remember why. I'll try and switch it back and see if I run into that issue again